### ISSUE TITLE
softmax(x) and logsoftmax(x) update their arguments

### ISSUE THREAD

**Issue by dfdx** [JuliaGPU/CUDA.jl#592]
**Describe the bug**

Unlike NNlib versions, `softmax(x)` and `logsoftmax(x)` in CUDA.jl modify their arguments in-place. 

**To reproduce**

The simplest way to confirm it is to look at [the code](https://github.com/JuliaGPU/CUDA.jl/blob/master/lib/cudnn/nnlib.jl#L48-L58), but for the sake of completeness here's the minimal reproducible example:

```julia
using CUDA
using NNlib

x = cu(rand(5, 10))
original_x = copy(x)
softmax(x)
@assert x == original_x
```

**Possible fix**

It should be easy to fix `softmax()`, `logsoftmax()`, `∇softmax()` and `∇logsoftmax()` by allocating a new array for the result, e.g.:
```julia
softmax(x::DenseCuArray{T}; dims=1) where T<:CUDNNFloat =
         softmax!(similar(x), x, dims=dims)
```
Right now I'm struggling with another issue stopping me from creating a PR, but I plan to come back to it later.

**Comment by denizyuret** [comment]
I am fixing this and other inconsistencies at https://github.com/JuliaGPU/CUDA.jl/pull/523

**Comment by CarloLucibello** [comment]
@maleadt would you take a fix to this for backporting on the julia 1.5 compatible branch? This unless #523 is meant to be backported

**Comment by maleadt** [comment]
> @maleadt would you take a fix to this for backporting on the julia 1.5 compatible branch?

Sure. The new CUDNN interface is unlikely to be backported.

**Comment by denizyuret** [comment]
@maleadt -- I don't think there is much backport work to do for the new cudnn interface, if you let me know what branch, I can check to make sure everything works for 1.5.

**Comment by maleadt** [comment]
The branch is called `compat`, I'll make a PR with the CUDNN changes for you and e.g. @DhairyaLGandhi or @CarloLucibello to verify works properly.

**Comment by DhairyaLGandhi** [comment]
Thanks, that's helpful

**Comment by jeremiedb** [comment]
@denizyuret I think this issue hasn't been solved at the moment (at least on latest CUDA 2.6.3 release): 

```julia
using CUDA
using Flux

julia> x = CuArray([-1.0, 0.0, 1.0])
3-element CuArray{Float64, 1}:
 -1.0
  0.0
  1.0

julia> softmax(x)
3-element CuArray{Float64, 1}:
 0.09003057317038046
 0.24472847105479767
 0.6652409557748219

julia> x
3-element CuArray{Float64, 1}:
 0.09003057317038046
 0.24472847105479767
 0.6652409557748219
```

Could it be reopened so as to keep track of it?

**Comment by denizyuret** [comment]
I think it is solved in master but not released yet.
